File: services/backend/app/core/google_auth.py
"""
core/google_auth.py
-------------------
Google OAuth 2.0 ID token verification for "Sign in with Google".

Accepts one or more Web client IDs (token `aud` claim):
  - GOOGLE_CLIENT_ID          (single ID, or comma-separated list)
  - GOOGLE_CLIENT_IDS         (optional extra comma-separated IDs)

Website GIS and native Android may use different Web client IDs
(e.g. website project vs Firebase project). Both must be listed.
"""

import logging
from typing import Any

from google.auth.transport import requests as google_requests
from google.oauth2 import id_token

from app.core.config import settings

logger = logging.getLogger(__name__)

# Known clients used by this project (always accepted if present in token).
# Safe defaults so mobile Firebase tokens work even if Azure only has the
# website GOOGLE_CLIENT_ID set. Override / extend via env in production.
_DEFAULT_EXTRA_CLIENT_IDS = (
    # Website GIS web client
    "596375721075-itbl6d2i44kekhniujmmm0g8jovoc9i6.apps.googleusercontent.com",
    # Firebase web client (native Android serverClientId)
    "609875913005-2od36vgq10osdp1ajohibcap37jab4ho.apps.googleusercontent.com",
)


def _split_ids(raw: str | None) -> list[str]:
    if not raw:
        return []
    out: list[str] = []
    for part in str(raw).replace(";", ",").split(","):
        p = part.strip().strip('"').strip("'")
        if p and p not in out:
            out.append(p)
    return out


def _allowed_client_ids() -> list[str]:
    """All Web client IDs that may appear as the ID-token audience."""
    ids: list[str] = []
    for source in (
        getattr(settings, "GOOGLE_CLIENT_ID", None),
        getattr(settings, "GOOGLE_CLIENT_IDS", None),
    ):
        for cid in _split_ids(source):
            if cid not in ids:
                ids.append(cid)
    # Always allow project-known clients so mobile/web don't drift
    for cid in _DEFAULT_EXTRA_CLIENT_IDS:
        if cid not in ids:
            ids.append(cid)
    return ids


def verify_google_token(token: str) -> dict[str, Any]:
    """
    Verify a Google ID token and return the payload.

    Raises:
        ValueError: if the token is invalid or audience doesn't match.
    """
    request = google_requests.Request()
    allowed = _allowed_client_ids()

    try:
        if not allowed:
            logger.warning(
                "GOOGLE_CLIENT_ID not set. Verifying without audience check (insecure for prod)."
            )
            return id_token.verify_oauth2_token(token, request)

        last_error: Exception | None = None
        for audience in allowed:
            try:
                return id_token.verify_oauth2_token(token, request, audience=audience)
            except ValueError as e:
                last_error = e
                logger.info(
                    "Google token not valid for audience %s…: %s",
                    audience[:24],
                    e,
                )

        # Decode without audience only to report what aud was (still reject if none matched)
        try:
            unverified = id_token.verify_oauth2_token(token, request)
            got_aud = unverified.get("aud")
        except Exception:
            got_aud = None

        msg = (
            f"Invalid Google token: Token has wrong audience {got_aud}, "
            f"expected one of {allowed}"
        )
        logger.warning("%s (last library error: %s)", msg, last_error)
        raise ValueError(msg)
    except ValueError:
        raise
    except Exception as e:
        logger.error("Error verifying Google token: %s", e)
        raise ValueError("Failed to verify Google token") from e

File: services/frontend/web/src/services/googleAuth.js
/**
 * googleAuth.js
 * -------------
 * Google Sign-In for web (GIS) and native Capacitor (Credential Manager).
 *
 * Native Android must use:
 *  - package name registered in Firebase / Google Cloud: com.company.smart_agri
 *  - SHA-1 of the APK signing cert (debug.keystore from packages/mobile)
 *  - Web client ID from the SAME Google project as the Android OAuth client
 *    so the ID token audience is accepted by the backend.
 */

import { Capacitor } from "@capacitor/core";
import { GoogleSignIn } from "@capawesome/capacitor-google-sign-in";

/**
 * Website GIS client (project 596375…).
 * Keep for browser "Sign in with Google" button.
 */
export const GOOGLE_WEB_CLIENT_ID =
  import.meta.env.VITE_GOOGLE_CLIENT_ID ||
  "596375721075-itbl6d2i44kekhniujmmm0g8jovoc9i6.apps.googleusercontent.com";

/**
 * Firebase / Android-linked Web client (project 609875…).
 * Used as serverClientId for native ID tokens so audience matches
 * the OAuth clients tied to package com.company.smart_agri.
 * Backend must list this in GOOGLE_CLIENT_ID or GOOGLE_CLIENT_IDS.
 */
export const GOOGLE_NATIVE_WEB_CLIENT_ID =
  import.meta.env.VITE_GOOGLE_NATIVE_CLIENT_ID ||
  "609875913005-2od36vgq10osdp1ajohibcap37jab4ho.apps.googleusercontent.com";

const NATIVE_PACKAGE = "com.company.smart_agri";

let nativeInitialized = false;

export function isNativePlatform() {
  return Capacitor.isNativePlatform();
}

export async function initNativeGoogleSignIn() {
  if (!isNativePlatform() || nativeInitialized) return;
  if (!GOOGLE_NATIVE_WEB_CLIENT_ID) {
    throw new Error("Native Google Client ID is not configured.");
  }
  await GoogleSignIn.initialize({
    // Must be a *Web* client ID from the Firebase Google project
    clientId: GOOGLE_NATIVE_WEB_CLIENT_ID,
  });
  nativeInitialized = true;
}

/**
 * @returns {Promise<string>} Google ID token for POST /auth/google
 */
export async function signInWithGoogleNative() {
  await initNativeGoogleSignIn();
  try {
    const result = await GoogleSignIn.signIn();
    if (!result?.idToken) {
      throw new Error(
        "Google did not return an ID token. Ensure Azure accepts the Firebase web client ID."
      );
    }
    return result.idToken;
  } catch (err) {
    const code = err?.code || err?.errorMessage || "";
    const msg = String(err?.message || err || "");

    if (code === "SIGN_IN_CANCELED" || /cancel/i.test(msg)) {
      const e = new Error("Google sign-in was cancelled.");
      e.code = "SIGN_IN_CANCELED";
      throw e;
    }

    if (
      /DEVELOPER_ERROR|ApiException:\s*10|\b10\b|mismatch|SHA|package|audience|client.?id/i.test(
        `${code} ${msg}`
      )
    ) {
      throw new Error(
        `Google Sign-In misconfigured. Package must be ${NATIVE_PACKAGE} ` +
          "and SHA-1 must match Firebase (5E:8F:16:06:…). " +
          "Also set Azure GOOGLE_CLIENT_IDS to include the Firebase web client ID."
      );
    }

    throw new Error(msg || "Google sign-in failed.");
  }
}
